AppBuilder/platform/views/viewComponent/ABViewOrgChartComponent.js

```diff
@@ -0,0 +1,224 @@
+
+      this.ready();
+   }
+
+   async onShow() {
+      super.onShow();
+
+      this.busy();
+      await this.loadOrgChartJs();
+      await this.pullData();
+      this.displayOrgChart();
+      this.ready();
+   }
+
+   async displayOrgChart() {
+      const baseView = this.view;
+      const chartData = this.AB.cloneDeep(this.chartData);
+
+      const orgchart = new this.OrgChart({
+         data: chartData,
+         direction: baseView.settings.direction,
+         // depth: baseView.settings.depth,
+         pan: baseView.settings.pan,
+         zoom: baseView.settings.zoom,
+         // visibleLevel: baseView.settings.visibleLevel,
+
+         exportButton: baseView.settings.export,
+         exportFilename: baseView.settings.exportFilename,
+
+         // ajaxURLs: {
+         //    children: function (nodeData) {
+         //       console.info("nodeData: ", nodeData);
+         //       return null;
+         //    },
+         // },
+         nodeContent: "description",
+      });
+
+      const chartDom = document.querySelector(`#${this.ids.chartDom}`);
+      if (chartDom) {
+         chartDom.textContent = "";
+         chartDom.innerHTML = "";
+         chartDom.appendChild(orgchart);
+      }
+
+      setTimeout(() => {
+         this._setColor();
+      }, 1);
+   }
+
+   async pullData() {
+      const view = this.view;
+      const dc = view.datacollection;
+      const cursor = dc?.getCursor();
+      if (!cursor) return null;
+
+      const valueFields = view.valueFields();
+      // const descriptionField = view.descriptionField?.();
+
+      const chartData = this.chartData;
+      chartData.name = dc?.datasource?.displayData(cursor) ?? "";
+      chartData.description = "";
+      // description:
+      //    descriptionField?.format?.(f) ??
+      //    f[descriptionField?.columnName] ??
+      //    "",
+      chartData._rawData = cursor;
+
+      let parentChartData = [chartData];
+      let currChildren;
+
+      valueFields.forEach((field) => {
+         let _tempParentChartData = [];
+
+         parentChartData.forEach(async (chartItem) => {
+            if (!chartItem) return;
+
+            const rawData = chartItem?._rawData;
+            currChildren = rawData?.[field?.relationName()];
+
+            // Pull data from the server
+            if (!currChildren) {
+               const objLink = field.object;
+               const where = {
+                  glue: "and",
+                  rules: [],
+               };
+               where.rules.push({
+                  key: objLink.PK(),
+                  rule: "equals",
+                  value: rawData[objLink.PK()],
+               });
+               const returnData = await objLink
+                  .model()
+                  .findAll({ where, populate: true });
+               chartItem._rawData = returnData?.data[0];
+               currChildren = chartItem._rawData?.[field?.relationName()];
+
+               this.displayOrgChart();
```

Comment:
this `displayOrgChart()` is inside 2 `.forEach()` statements.

Wouldn't this be better off after all the loops are done?

Reply:
The first loop iterates through each field.
The other loop processes the data rows.